### ISSUE TITLE
There is a Security Flaw

### ISSUE THREAD

**Issue by 2shrestha22** [Ashwin1002/ConnectIPS-Flutter#1]
Thanks for the package. But there is a major security flaw. You should not put any private key or secret in the app. They are easy to extract. While you suggest to add private key in the assets it will lead to security issue.

**Comment by Ashwin1002** [comment]
I am currently working on an alternative to replace this security concern. Any feedback about what can be done to remove this flaw?

**Comment by 2shrestha22** [comment]
You need to move the token signing functionality to the backend. Without backend it will never be secure.

**Comment by Ashwin1002** [comment]
I just updated the code and readme file to prompt the user to get the secret key from the server.

**Comment by 2shrestha22** [comment]
@Ashwin1002 you didn't understand. Storing secrets in app or retrieving it from API is same and there is no security fix. API requests can be easily monitored. You send the data to the backend and backend sends you signed token without needing to fetch the secret.

**Comment by Ashwin1002** [comment]
Thank you for your concern. Please look into the npi documents, and
speaking with the npi team, the token that is used for signing is done from
the client and both their server sides, and they check if the token
generated from both sides is the same. It is the connect IPS that has made
the merchant login page validate that way. From your perspective, you want
me to drop the package as there are no alternatives other than fetching the
API through the backend. What about a serverless backend? If I am using
tools like Firebase or Supabase, then how does the user integrate this
payment gateway? I think giving the user choice might be the best in this
case.

**Comment by 2shrestha22** [comment]
- I don't have access to NIP documents.
- What is the reason of encrypting? If NIP team says it's okay to not secure that key that you are calling it private. May be it's a public key. See below about public and private key.
- I am not saying you to drop this package.
- Serverless backend alwasy have serverless functions where you can write APIs or do whatever you want.
- It's not about right choice. It's about proper implementation.

Please read this: https://stackoverflow.com/a/454069

**Comment by Ashwin1002** [comment]
Thank you for your wonderful information. I didn't have an idea about the
actual functionality of the private keys. With little backend knowledge, I
just followed the document steps and also got help from the nchl team to
implement this in my project and just shared how I achieved this in
Flutter. I just documented the steps and the problems that I faced when
implementing the ConnectIPS gateway. If this package has a flaw, then I
would recommend other developers about the possible risks. Thank you for
raising this issue.
